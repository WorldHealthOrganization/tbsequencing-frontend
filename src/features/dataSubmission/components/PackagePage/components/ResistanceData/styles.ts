import { css } from '@emotion/react';
import { fontSizes } from '../../../../../../components/typography/fontSizes';
import appColors from '../../../../../../styles/colors';
import { appZIndex } from '../../../../../../styles/appZIndex';

export const container = css({
  width: '60%',
  boxShadow: `0px 0px 4px ${appColors.tints.whoBlue.lighter}`,
});

export const header = css({
  fontWeight: 700,
  paddingLeft: 24,
  paddingTop: 16,
});

export const sxTabStyles = {
  fontSize: fontSizes.primary,
  lineHeight: '24px',
  paddingTop: '11px',
  paddingBottom: '11px',
  color: appColors.tints.whoBlue.light,
  '&.Mui-selected': {
    color: appColors.secondary.whoBlue,
  },
  textTransform: 'capitalize',
};

export const sxTabsStyles = {
  '& .MuiTabs-indicator': {
    display: 'flex',
    justifyContent: 'center',
    backgroundColor: appColors.secondary.whoBlue,
    height: 2,
    zIndex: appZIndex.appBar,
  },
};

export const tabsWrapper = css({
  position: 'relative',
});

export const fakeBorderBottom = css({
  height: 2,
  backgroundColor: appColors.tints.whoBlue.lightest,
  position: 'absolute',
  bottom: 0,
  left: 0,
  right: 0,
});

export const loaderWrapper = css({
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  paddingTop: 24,
});

export const samplesContent = css({
  position: 'relative',
  minHeight: '35px',
});
