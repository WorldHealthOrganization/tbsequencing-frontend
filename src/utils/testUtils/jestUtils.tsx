import React, { ReactElement } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { reducer } from '../../app/store';

const getWrapper = (
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
  } = {},
) => ({ children }: { children: ReactElement }) => (
  <Provider store={store}>{children}</Provider>
);

function render(
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
    ...renderOptions
  } = {},
) {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

const getTreeJSONWithRedux = (ui: ReactElement, {
  // @ts-ignore
  preloadedState,
  store = configureStore({ reducer, preloadedState }),
} = {}) => {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  const tree = renderer
    .create(<Wrapper>{ui}</Wrapper>)
    .toJSON();

  return tree;
};

export const wrappedInRouter = (
  component: ReactElement,
) => <BrowserRouter>{component}</BrowserRouter>;

export * from '@testing-library/react';
export { render, getTreeJSONWithRedux };
