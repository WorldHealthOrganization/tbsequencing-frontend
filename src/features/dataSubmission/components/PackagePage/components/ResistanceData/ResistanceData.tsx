/** @jsxImportSource @emotion/react */
import React from 'react';
import {
  Tab, Tabs, Alert,
} from '@mui/material';
import * as styles from './styles';
import type { ITab } from '../../PackagePage';
import {
  ResistanceSampleType,
  GetSamplesByFileIdResponse,
} from '../../../../../../services/submissionApi/models';
import RemoveResistanceFileRow from './components/RemoveResistanceFileRow';
import DataSubmissionSampleRow from '../DataSubmissionSampleRow';
import ResistanceDataTable from './components/ResistanceDataTable';
import ActivityIndicator from '../../../../../../components/ActivityIndicator';
import H3 from '../../../../../../components/typography/H3';
import { getResistanceSamplesDataByValidity, getUnprocessedResistanceSamples } from './utils/dataUtils';
import { IHttpError } from '../../../../../../models/HttpError';

interface Props {
  activeTab: ResistanceSampleType
  tabs: ITab[];
  handleTabChange: (event: React.SyntheticEvent, newValue: ResistanceSampleType) => void;
  samplesData: GetSamplesByFileIdResponse;
  onRemoveFileClick: () => void;
  onRemoveErrorClick: () => void;
  isTabFileUploadError: boolean;
  errorFileName: string;
  isLoading: boolean;
  controlsDisabled: boolean;
  packageId: number;
  resistanceFileLabel: string;
  fileError: IHttpError;
  notUsedColumns: string[]
}

export const ResistanceData = ({
  tabs, activeTab, handleTabChange, samplesData, isLoading,
  onRemoveFileClick, isTabFileUploadError, errorFileName,
  onRemoveErrorClick, controlsDisabled, packageId,
  resistanceFileLabel, fileError, notUsedColumns,
}: Props) => {
  const renderSamplesContent = () => {
    if (isLoading) {
      return <ActivityIndicator size={20} centered />;
    }

    const validSamples = getResistanceSamplesDataByValidity(samplesData, true);
    const invalidSamples = getResistanceSamplesDataByValidity(samplesData, false);
    const unprocessedSamples = getUnprocessedResistanceSamples(samplesData);
    const shouldRenderValidTable = validSamples.length > 0;
    const shouldRenderInvalidTable = invalidSamples.length > 0;
    const shouldRenderUnprocessedTable = unprocessedSamples.length > 0;
    const shouldRenderFileLink = samplesData.length > 0;
    const errorMsg = fileError?.data?.errors[0].detail;

    return (
      <>
        {isTabFileUploadError && (
          <DataSubmissionSampleRow
            controlDisabled={controlsDisabled}
            title={errorFileName}
            isError={isTabFileUploadError}
            onRemoveClick={onRemoveErrorClick}
            description={errorMsg}
          />

        )}
        <>
          {shouldRenderFileLink && (
            <RemoveResistanceFileRow
              controlsDisabled={controlsDisabled}
              onRemoveClick={onRemoveFileClick}
              fileName={resistanceFileLabel}
            />
          )}
          {notUsedColumns?.length > 0 && (
            <Alert variant="filled" severity="warning">
              The following columns are not used:
              &nbsp;
              {notUsedColumns.join(', ')}
            </Alert>
          )}
          {shouldRenderUnprocessedTable && (
            <ResistanceDataTable
              activeTab={activeTab}
              controlsDisabled={controlsDisabled}
              packageId={packageId}
              samplesData={unprocessedSamples}
              title="Unprocessed"
            />
          ) }
          {shouldRenderInvalidTable && (
            <ResistanceDataTable
              activeTab={activeTab}
              controlsDisabled={controlsDisabled}
              packageId={packageId}
              samplesData={invalidSamples}
              title="Have Issues"
            />
          )}
          {shouldRenderValidTable && (
            <ResistanceDataTable
              activeTab={activeTab}
              controlsDisabled={controlsDisabled}
              packageId={packageId}
              samplesData={validSamples}
              title="Successfully Processed"
            />
          )}
        </>
      </>
    );
  };

  return (
    <div css={styles.container}>
      <H3 style={styles.header}>Resistance Data</H3>
      <div css={styles.tabsWrapper}>
        <Tabs
          sx={styles.sxTabsStyles}
          variant="fullWidth"
          value={activeTab}
          onChange={handleTabChange}
        >
          {tabs.map((tab) => (
            <Tab {...tab} key={tab.value} sx={styles.sxTabStyles} />
          ))}
        </Tabs>
        <div css={styles.fakeBorderBottom} />
      </div>
      <div css={styles.samplesContent}>
        {renderSamplesContent()}
      </div>
    </div>
  );
};

export default ResistanceData;
