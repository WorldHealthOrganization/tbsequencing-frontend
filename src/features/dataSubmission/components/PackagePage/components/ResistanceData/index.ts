import ResistanceData from './ResistanceData';

export default ResistanceData;
