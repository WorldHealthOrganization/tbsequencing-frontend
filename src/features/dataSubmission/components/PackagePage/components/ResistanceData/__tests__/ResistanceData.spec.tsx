import React from 'react';
import { getTreeJSONWithRedux, render } from '../../../../../../../utils/testUtils/jestUtils';
import ResistanceData from '../ResistanceData';

describe('ResistanceData component test', () => {
  const mockOnClick = jest.fn();

  test('renders ResistanceData component without error', () => {
    render(
      <ResistanceData
        isLoading={false}
        handleTabChange={mockOnClick}
        onRemoveErrorClick={mockOnClick}
        activeTab="PDST"
        tabs={[{
          value: 'PDST',
          label: 'mock',
        }]}
        samplesData={[]}
        isTabFileUploadError={false}
        errorFileName=""
        onRemoveFileClick={mockOnClick}
      />,
    );
  });

  it('Should render correctly', () => {
    const tree = getTreeJSONWithRedux(
      <ResistanceData
        isLoading={false}
        handleTabChange={mockOnClick}
        onRemoveErrorClick={mockOnClick}
        activeTab="PDST"
        tabs={[{
          value: 'PDST',
          label: 'mock',
        }]}
        samplesData={[]}
        isTabFileUploadError={false}
        errorFileName=""
        onRemoveFileClick={mockOnClick}
      />,
    );

    expect(tree).toMatchSnapshot();
  });
});
