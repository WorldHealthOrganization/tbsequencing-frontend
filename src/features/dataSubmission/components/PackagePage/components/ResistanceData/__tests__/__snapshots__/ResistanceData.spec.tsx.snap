// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResistanceData component test Should render correctly 1`] = `
<div
  className="css-t8y96o-ResistanceData"
>
  <p
    className="MuiTypography-root MuiTypography-body1 css-ve7s52-MuiTypography-root-BaseText"
  >
    Resistance Data
  </p>
  <div
    className="css-wh98kw-ResistanceData"
  >
    <div
      className="MuiTabs-root css-1h164f2-MuiTabs-root"
    >
      <div
        className="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
        onScroll={[Function]}
        style={
          Object {
            "marginBottom": -0,
            "overflow": "hidden",
          }
        }
      >
        <div
          aria-orientation={null}
          className="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
          onKeyDown={[Function]}
          role="tablist"
        >
          <button
            aria-selected={true}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary MuiTab-fullWidth Mui-selected css-109t08r-MuiButtonBase-root-MuiTab-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={0}
            type="button"
          >
            mock
            <span
              className="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
              style={Object {}}
            />
          </button>
        </div>
      </div>
    </div>
    <div
      className="css-1d4gvh7-ResistanceData"
    />
  </div>
  <div
    className="css-140p7hc-ResistanceData"
  />
</div>
`;
